src/shared/settingsConfiguration.ts

```diff
@@ -5,231 +5,590 @@
+const settingsProps = packageJson.contributes.configuration.properties
+
+type SettingsProps = typeof settingsProps
+
+type Split<T, S extends string> = T extends `${infer L}${S}${infer R}` ? [L, ...Split<R, S>] : [T]
+type Pop<T> = T extends [...infer R, infer _] ? R : never
+type Intersection<T> = (T extends any ? (_: T) => any : never) extends (_: infer U) => any ? U : never
+
+type FromParts<T, K> = K extends [infer U, ...infer R]
+    ? [U, R] extends [string, string[]]
+        ? { [P in U & string]: FromParts<T, R> }
+        : T
+    : T
+
+type Format<T> = { [P in keyof T]: FromParts<TypeConstructor, Split<P, '.'>> }[keyof T]
+type Config = Intersection<Format<SettingsProps>>
+
+type Join<T extends string[], S extends string> = T['length'] extends 1
+    ? T[0]
+    : T extends [infer L, ...infer R]
+    ? L extends string
+        ? R extends string[]
+            ? `${L}${S}${Join<R, S>}`
+            : ''
+        : ''
+    : never
+
+type Select<T, K> = K extends [infer L, ...infer R]
+    ? L extends keyof T
+        ? R['length'] extends 0
+            ? T[L]
+            : Select<T[L], R>
+        : never
+    : never
+
+type Sections = { [P in keyof SettingsProps as Join<Pop<Split<P, '.'>>, '.'>]: Select<Config, Pop<Split<P, '.'>>> }
+
+/**
+ * Creates a class for manipulating specific sections of settings specified in `package.json`.
+ *
+ * A 'section' is the combined namespaces of a key, e.g. `aws.foo.bar` is within the section `aws.foo`.
+ * From a section, callers can specify which keys they want to use as well as how it should be parsed.
+ * The returned class will be a valid implementation of {@link TypedSettings}. Declared settings keys
+ * are validated against `package.json` to ensure they exist.
+ *
+ * ### Examples
+ * #### Pass-through:
+ * ```
+ * export class CloudWatchLogsSettings extends fromPackage('aws.cloudWatchLogs', { limit: Number }) {}
+ *
+ * const settings = new CloudWatchLogsSettings()
+ * const limit = settings.get('limit', 1000)
+ * ```
+ *
+ * #### Extending:
+ * ```
+ * export class TelemetryConfig extends fromPackage('aws', { telemetry: Boolean }) {
+ *     public isEnabled(): boolean {
+ *         try {
+ *             return this.get(TELEMETRY_KEY, TELEMETRY_SETTING_DEFAULT)
+ *         } catch (error) {
+ *              vscode.window.showErrorMessage(
+ *                  localize(
+ *                      'AWS.message.error.settings.telemetry.invalid_type',
+ *                      'The aws.telemetry value must be a boolean'
+ *                  )
+ *              )
+ *
+ *              return TELEMETRY_SETTING_DEFAULT
+ *         }
+ *     }
+ * }
+ * ```
+ *
+ * @param section Section used to create the class
+ * @param descriptor A {@link TypeDescriptor} that maps keys to a concrete type
+ *
+ * @returns A class that can be used as-is or inherited from
+ */
+export function fromPackage<T extends TypeDescriptor & Partial<Sections[K]>, K extends keyof Sections>(
```

Comment:
"package" was throwing me off, it sounds like it takes any NPM package or something. I wonder if "package json" is more intuitive.
```suggestion
export function fromPackageJson<T extends TypeDescriptor & Partial<Sections[K]>, K extends keyof Sections>(
```